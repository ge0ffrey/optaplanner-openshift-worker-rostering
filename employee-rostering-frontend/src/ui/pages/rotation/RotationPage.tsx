/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from "react";
import Spot from "domain/Spot";
import { AppState } from "store/types";
import { spotSelectors } from "store/spot";
import { connect } from 'react-redux';
import moment from 'moment';
import { Level, LevelItem, Button, Text, Pagination, ButtonVariant } from "@patternfly/react-core";
import { EventProps } from 'react-big-calendar'
import { modulo } from 'util/MathUtils';
import TypeaheadSelectInput from "ui/components/TypeaheadSelectInput";
import { alert } from "store/alert";
import RosterState from "domain/RosterState";

import ShiftTemplate from "domain/ShiftTemplate";
import { shiftTemplateSelectors, shiftTemplateOperations } from "store/rotation";
import { WithTranslation, withTranslation, useTranslation } from "react-i18next";
import EditShiftTemplateModal from "./EditShiftTemplateModal";
import { EditIcon, TrashIcon } from "@patternfly/react-icons";
import Schedule from "ui/components/calendar/Schedule";

interface StateProps {
  isLoading: boolean;
  spotList: Spot[];
  spotIdToShiftTemplateListMap: Map<number, ShiftTemplate[]>;
  rosterState: RosterState | null;
}
  
const mapStateToProps = (state: AppState): StateProps => ({
  isLoading: state.shiftTemplateList.isLoading,
  spotList: spotSelectors.getSpotList(state),
  spotIdToShiftTemplateListMap: shiftTemplateSelectors.getShiftTemplateList(state)
    .reduce((prev, curr) => {
      const old = prev.get(curr.spot.id as number)? prev.get(curr.spot.id as number) as ShiftTemplate[] : [];
      const templatesToAdd: ShiftTemplate[] =
      (state.rosterState.rosterState !== null &&
        moment.duration(curr.durationBetweenRotationStartAndTemplateStart)
          .add(curr.shiftTemplateDuration).asDays() >= state.rosterState.rosterState.rotationLength)?
        [curr, {
          ...curr,
          durationBetweenRotationStartAndTemplateStart:
           moment.duration(-state.rosterState.rosterState.rotationLength, "days")
             .add(curr.durationBetweenRotationStartAndTemplateStart)
        }] : [curr];
      return prev.set(curr.spot.id as number, old.concat(templatesToAdd));
    },
    new Map<number, ShiftTemplate[]>()),
  rosterState: state.rosterState.rosterState
}); 
  
export interface DispatchProps {
  addShiftTemplate: typeof shiftTemplateOperations.addShiftTemplate;
  removeShiftTemplate: typeof shiftTemplateOperations.removeShiftTemplate;
  updateShiftTemplate: typeof shiftTemplateOperations.updateShiftTemplate;
  showInfoMessage: typeof alert.showInfoMessage;
}
  
const mapDispatchToProps: DispatchProps = {
  addShiftTemplate: shiftTemplateOperations.addShiftTemplate,
  removeShiftTemplate: shiftTemplateOperations.removeShiftTemplate,
  updateShiftTemplate: shiftTemplateOperations.updateShiftTemplate,
  showInfoMessage: alert.showInfoMessage
}
  
export type Props = StateProps & DispatchProps;
interface State {
  shownSpot: Spot|null;
  isCreatingOrEditingShiftTemplate: boolean;
  selectedShiftTemplate?: ShiftTemplate;
  weekNumber: number;
}

export const baseDate = moment("2018-01-01T00:00").startOf('week').toDate();

const ShiftTemplatePopoverHeader: React.FC<{
  shiftTemplate: ShiftTemplate;
  rotationLength: number;
  onEdit: (shift: ShiftTemplate) => void;
  onDelete: (shift: ShiftTemplate) => void;
}> = props =>{
  const { t } = useTranslation();
  const durationBetweenRotationStartAndEnd = moment
    .duration(props.shiftTemplate.durationBetweenRotationStartAndTemplateStart)
    .add(props.shiftTemplate.shiftTemplateDuration);
  return (
    <span> 
      <Text> 
        {t("shiftTemplate", {
          spot: props.shiftTemplate.spot.name,
          rotationEmployee: props.shiftTemplate.rotationEmployee? props.shiftTemplate.rotationEmployee.name
            : "Unassigned",
          dayStart: Math.floor(modulo(
            props.shiftTemplate.durationBetweenRotationStartAndTemplateStart.asDays(),
            props.rotationLength)) + 1,
          startTime: moment("2018-01-01")
            .add(props.shiftTemplate.durationBetweenRotationStartAndTemplateStart).format("LT"),
          dayEnd: Math.floor(modulo(durationBetweenRotationStartAndEnd.asDays(),
            props.rotationLength)) + 1,
          endTime: moment("2018-01-01")
            .add(durationBetweenRotationStartAndEnd).format("LT")
        })}

      </Text>
      <Button
        onClick={() => props.onEdit(props.shiftTemplate)}
        variant={ButtonVariant.link}
      >
        <EditIcon />
      </Button>
      <Button
        onClick={() => props.onDelete(props.shiftTemplate)}
        variant={ButtonVariant.link}
      >
        <TrashIcon />
      </Button>
    </span>
  )}

const ShiftTemplatePopoverBody: React.FC = () => <></>;

const ShiftTemplateEvent: React.FC<EventProps<ShiftTemplate>> = (props) => {

  return (
    <span
      data-tip
      data-for={String(props.event.id)}
      style={{
        display: "flex",
        height: "100%",
        width: "100%"
      }}
    >
      {props.title}
    </span>
  )
};

export class RotationPage extends React.Component<Props & WithTranslation, State> {
  constructor(props: Props & WithTranslation) {
    super(props);
    this.addShiftTemplate = this.addShiftTemplate.bind(this);
    this.updateShiftTemplate = this.updateShiftTemplate.bind(this);
    this.deleteShiftTemplate = this.deleteShiftTemplate.bind(this);

    const shownSpot = (props.spotList.length > 0)? props.spotList[0] : null;
    this.state = {
      isCreatingOrEditingShiftTemplate: false,
      weekNumber: 0,
      shownSpot
    };
  }

  componentDidUpdate() {
    const shownSpot = this.state.shownSpot;
    if (this.props.spotList.length > 0 && ((shownSpot !== null && 
      this.props.spotList.find((spot) => spot.id === shownSpot.id) === undefined) ||
      (shownSpot === null))
    ) {
      // eslint-disable-next-line react/no-did-update-set-state
      this.setState({
        shownSpot: this.props.spotList[0]
      });
    }
  }

  addShiftTemplate(addedShiftTemplate: ShiftTemplate) {
    this.props.addShiftTemplate(addedShiftTemplate);
  }

  updateShiftTemplate(updatedShiftTemplate: ShiftTemplate) {
    this.props.updateShiftTemplate(updatedShiftTemplate);
  }


  deleteShiftTemplate(deletedShiftTemplate: ShiftTemplate) {
    this.props.removeShiftTemplate(deletedShiftTemplate);
  }

  render() {
    if (this.props.rosterState === null || this.props.isLoading || this.props.spotList.length <= 0 || 
      this.state.shownSpot === null || !this.props.tReady) {
      return <div />;
    }

    const startDate = moment(baseDate).add(this.state.weekNumber, "weeks").toDate();
    const endDate = moment(startDate).add(1, "week").toDate();
    const spot = this.state.shownSpot;
    const events: { shiftTemplate: ShiftTemplate; start: Date; end: Date }[] = [];

    (this.props.spotIdToShiftTemplateListMap.get(this.state.shownSpot.id as number) as ShiftTemplate[])
      .forEach(st => {
        const startHours = st.shiftTemplateDuration.hours();
        const startMinutes = st.shiftTemplateDuration.minutes();
        const durationBetweenRotationStartAndTemplateStart = 
          moment.duration(st.durationBetweenRotationStartAndTemplateStart);

        if (startHours === 0 && startMinutes === 0) { 
          durationBetweenRotationStartAndTemplateStart.add(1, "ms");
        }


        const durationBetweenRotationStartAndEnd = moment
          .duration(st.durationBetweenRotationStartAndTemplateStart).add(st.shiftTemplateDuration);
        const endHours = durationBetweenRotationStartAndEnd.hours();
        const endMinutes = durationBetweenRotationStartAndEnd.minutes();
        const shiftTemplateDuration = moment.duration(st.shiftTemplateDuration);
        if (endHours === 0 && endMinutes=== 0) {
          shiftTemplateDuration.subtract(1, "ms");
        }
        events.push({
          shiftTemplate: st,
          start:  moment(baseDate).add(durationBetweenRotationStartAndTemplateStart).toDate(),
          end: moment(baseDate)
            .add(durationBetweenRotationStartAndTemplateStart)
            .add(shiftTemplateDuration).toDate()
        });
      });

    return (
      <>
        <Level
          gutter="sm"
          style={{
            height: "60px",
            padding: "5px 5px 5px 5px",
            backgroundColor: "var(--pf-global--BackgroundColor--100)"
          }}
        >
          <LevelItem style={{display: "flex"}}>
            <TypeaheadSelectInput
              aria-label="Select Spot"
              emptyText="Select Spot"
              optionToStringMap={spot => spot.name}
              options={this.props.spotList}
              defaultValue={this.state.shownSpot}
              onChange={(newSpot) => {
                if (newSpot !== undefined) {
                  this.setState({ shownSpot: newSpot });
                }
              }}
            />
            <Pagination
              itemCount={Math.ceil(this.props.rosterState.rotationLength)}
              page={this.state.weekNumber + 1}
              onSetPage={(e, page) => {
                this.setState({ weekNumber: page - 1 });
              }}
              perPage={7}
              perPageOptions={[]}
              titles={{
                items: 'days',
                page: 'week',
                itemsPerPage: 'Week #',
                perPageSuffix: 'Week #',
                toFirstPage: 'Go to the first week',
                toPreviousPage: 'Go to the previous week',
                toLastPage: 'Go to the last week',
                toNextPage: 'Go to the next week',
                optionsToggle: 'Select',
                currPage: 'Current week',
                paginationTitle: 'Week Select'
              }}
            />
          </LevelItem>
          <LevelItem style={{display: "flex"}}>
            <Button
              style={{margin: "5px"}}
              aria-label="Create Shift Template"
              onClick={() => {
                this.setState({
                  selectedShiftTemplate: undefined,
                  isCreatingOrEditingShiftTemplate: true
                })
              }}
            >
              Create Shift Template
            </Button>
          </LevelItem>
        </Level>
        <EditShiftTemplateModal
          aria-label="Edit Shift Template"
          shiftTemplate={this.state.selectedShiftTemplate}
          isOpen={this.state.isCreatingOrEditingShiftTemplate}
          onSave={shiftTemplate => {
            if (this.state.selectedShiftTemplate !== undefined) {
              this.props.updateShiftTemplate(shiftTemplate);
            }
            else {
              this.props.addShiftTemplate(shiftTemplate);
            }
            this.setState({ selectedShiftTemplate: undefined, isCreatingOrEditingShiftTemplate: false });
          }}
          onDelete={shiftTemplate => {
            this.props.removeShiftTemplate(shiftTemplate);
            this.setState({ isCreatingOrEditingShiftTemplate: false });
          }}
          onClose={() => {
            this.setState({ 
              selectedShiftTemplate: undefined, 
              isCreatingOrEditingShiftTemplate: false
            });
          }}
        />
        <Schedule<{ shiftTemplate: ShiftTemplate; start: Date; end: Date }>
          key={spot.id}
          startDate={startDate}
          endDate={endDate}
          dateFormat={(date) =>
            this.props.t("rotationDay",
              {
                day: moment.duration(moment(date).diff(baseDate)).asDays() + 1
              })
          }
          events={events}
          titleAccessor={e => e.shiftTemplate.rotationEmployee?
            e.shiftTemplate.rotationEmployee.name : "Unassigned"}
          startAccessor={e => e.start}
          endAccessor={e => e.end}
          addEvent={
            (start,end) => {this.addShiftTemplate({
              tenantId: spot.tenantId,
              durationBetweenRotationStartAndTemplateStart: moment.duration(moment(start)
                .diff(baseDate)),
              shiftTemplateDuration: moment.duration(moment(end).diff(baseDate))
                .subtract(moment(start).diff(baseDate)),
              spot: spot,
              rotationEmployee: null,
            });}
          }
          eventStyle={() => ({})}
          dayStyle={() => ({})}
          wrapperStyle={() => ({})}
          popoverHeader={st => ShiftTemplatePopoverHeader({
            shiftTemplate: st.shiftTemplate,
            rotationLength: (this.props.rosterState as RosterState).rotationLength,
            onEdit: (shiftTemplate) => this.setState({
              selectedShiftTemplate: shiftTemplate,
              isCreatingOrEditingShiftTemplate: true
            }),
            onDelete: (shiftTemplate) => this.props.removeShiftTemplate(shiftTemplate)
          })
          }
          popoverBody={() => ShiftTemplatePopoverBody}
          eventComponent={(params) => ShiftTemplateEvent({
            ...params,
            event: params.event.shiftTemplate,
          })}
        />
      </>
    );
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(withTranslation()(RotationPage));