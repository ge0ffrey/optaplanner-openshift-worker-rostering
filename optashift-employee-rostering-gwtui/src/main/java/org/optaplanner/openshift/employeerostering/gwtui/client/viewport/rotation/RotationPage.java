package org.optaplanner.openshift.employeerostering.gwtui.client.viewport.rotation;

import javax.annotation.PostConstruct;
import javax.enterprise.event.Observes;
import javax.inject.Inject;

import elemental2.promise.Promise;
import org.jboss.errai.ui.client.local.api.elemental2.IsElement;
import org.jboss.errai.ui.shared.api.annotations.DataField;
import org.jboss.errai.ui.shared.api.annotations.Templated;
import org.optaplanner.openshift.employeerostering.gwtui.client.header.HeaderView;
import org.optaplanner.openshift.employeerostering.gwtui.client.pages.Page;
import org.optaplanner.openshift.employeerostering.gwtui.client.tenant.TenantStore;
import org.optaplanner.openshift.employeerostering.gwtui.client.viewport.CSSGlobalStyle;
import org.optaplanner.openshift.employeerostering.gwtui.client.viewport.CSSGlobalStyle.GridVariables;

@Templated
public class RotationPage implements IsElement,
                          Page {

    @Inject
    @DataField("viewport")
    private RotationPageViewport viewport;

    @Inject
    private RotationPageViewportBuilder viewportBuilder;

    @Inject
    private HeaderView headerView;

    @Inject
    private RotationToolbar toolbar;

    @Inject
    private CSSGlobalStyle cssGlobalStyle;

    @PostConstruct
    public void init() {
        cssGlobalStyle.setGridVariable(GridVariables.GRID_UNIT_SIZE, 10);
        cssGlobalStyle.setGridVariable(GridVariables.GRID_SOFT_LINE_INTERVAL, 4);
        cssGlobalStyle.setGridVariable(GridVariables.GRID_HARD_LINE_INTERVAL, 24);
        cssGlobalStyle.setGridVariable(GridVariables.GRID_HEADER_COLUMN_WIDTH, 120);
    }

    @Override
    public Promise<Void> onOpen() {
        headerView.addStickyElement(toolbar);
        return refresh();
    }

    public void onTenantChanged(@Observes final TenantStore.TenantChange tenant) {
        refresh();
    }

    private Promise<Void> refresh() {
        return viewportBuilder.buildRotationViewport(viewport);
    }
}
