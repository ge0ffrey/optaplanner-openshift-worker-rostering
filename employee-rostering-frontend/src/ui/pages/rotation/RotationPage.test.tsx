/*
 * Copyright 2019 Red Hat, Inc. and/or its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import * as React from 'react';
import Spot from 'domain/Spot';
import Employee from 'domain/Employee';
import { RotationPage, Props, baseDate } from './RotationPage';
import RosterState from 'domain/RosterState';
import moment from 'moment-timezone';
import ShiftTemplate from 'domain/ShiftTemplate';
import { useTranslation, WithTranslation } from 'react-i18next';

describe('Rotation Page', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render correctly when loaded', () => {
    const rotationPage = shallow(<RotationPage
      {...baseProps}
    />);
    expect(toJson(rotationPage)).toMatchSnapshot();
  });

  it('should render correctly when loading', () => {
    const rotationPage = shallow(<RotationPage
      {...baseProps}
      isLoading
      spotList={[]}
      spotIdToShiftTemplateListMap={new Map()}
    />);
    expect(toJson(rotationPage)).toMatchSnapshot();
  });

  it('should render correctly when creating a new shift template via button', () => {
    const rotationPage = shallow(<RotationPage
      {...baseProps}
    />);
    rotationPage.find('Button[aria-label="Create Shift Template"]').simulate("click");
    expect(toJson(rotationPage)).toMatchSnapshot();
  });

  it('should update weekNumber state when page change', () => {
    const rotationPage = shallow(<RotationPage
      {...baseProps}
    />);
    rotationPage.find('Pagination').simulate("setPage", null, 2);
    expect(rotationPage.state("weekNumber")).toEqual(1);
  });

  it('should update shownSpot on spot change', () => {
    const rotationPage = shallow(<RotationPage
      {...baseProps}
    />);
    const newSpot: Spot = {
      ...spot,
      id: 111,
      name: "New Spot"
    };

    const setStateSpy = jest.spyOn(rotationPage.instance(), "setState");
    rotationPage.find('TypeaheadSelectInput[aria-label="Select Spot"]').simulate("change", newSpot);
    // Bug in enzyme; wrapper.state is not updated after setState sometimes
    expect(setStateSpy).toBeCalledWith({ shownSpot: newSpot });
  });

  it('should not update shownSpot on spot change if invalid', () => {
    const rotationPage = shallow(<RotationPage
      {...baseProps}
    />);
    const setStateSpy = jest.spyOn(rotationPage.instance(), "setState");
    rotationPage.find('TypeaheadSelectInput[aria-label="Select Spot"]').simulate("change");
    expect(setStateSpy).not.toBeCalled();
  });

  it('should call addShift on addShift', () => {
    const rotationPage = new RotationPage(baseProps);
    const newShiftTemplate: ShiftTemplate = {
      ...shiftTemplate,
      id: undefined,
      version: undefined
    };
    rotationPage.addShiftTemplate(newShiftTemplate);
    expect(baseProps.addShiftTemplate).toBeCalled();
    expect(baseProps.addShiftTemplate).toBeCalledWith(newShiftTemplate);
  });

  it('should call updateShift on updateShift', () => {
    const rotationPage = new RotationPage(baseProps);
    const updatedShiftTemplate: ShiftTemplate = {
      ...shiftTemplate,
    };
    rotationPage.updateShiftTemplate(updatedShiftTemplate);
    expect(baseProps.updateShiftTemplate).toBeCalled();
    expect(baseProps.updateShiftTemplate).toBeCalledWith(updatedShiftTemplate);
  });

  it('should call removeShiftTemplate on deleteShiftTemplate', () => {
    const rotationPage = new RotationPage(baseProps);
    rotationPage.deleteShiftTemplate(shiftTemplate);
    expect(baseProps.removeShiftTemplate).toBeCalled();
    expect(baseProps.removeShiftTemplate).toBeCalledWith(shiftTemplate);
  });

  it('call deleteShift when the EditShiftModal delete a shift', () => {
    const rotationPage = shallow(<RotationPage
      {...baseProps}
    />);
    rotationPage.setState({
      selectedShift: shiftTemplate,
      isCreatingOrEditingShiftTemplate: true
    });
    rotationPage.find('Connect(Component)[aria-label="Edit Shift Template"]')
      .simulate("delete", shiftTemplate);
    expect(baseProps.removeShiftTemplate).toBeCalled();
    expect(baseProps.removeShiftTemplate).toBeCalledWith(shiftTemplate);
    expect(rotationPage.state("isCreatingOrEditingShiftTemplate")).toEqual(false);
  });

  it('call addShiftTemplate when the EditShiftTemplateModal add a new shift', () => {
    const rotationPage = shallow(<RotationPage
      {...baseProps}
    />);
    rotationPage.setState({
      selectedShift: null,
      isCreatingOrEditingShiftTemplate: true
    });
    rotationPage.find('Connect(Component)[aria-label="Edit Shift Template"]')
      .simulate("save", shiftTemplate);
    expect(baseProps.addShiftTemplate).toBeCalled();
    expect(baseProps.addShiftTemplate).toBeCalledWith(shiftTemplate);
    expect(rotationPage.state("isCreatingOrEditingShiftTemplate")).toEqual(false);
  });

  it('call updateShift when the EditShiftModal updates a shift', () => {
    const rotationPage = shallow(<RotationPage
      {...baseProps}
    />);
    rotationPage.setState({
      selectedShift: shiftTemplate,
      isCreatingOrEditingShiftTemplate: true
    });
    rotationPage.find('Connect(Component)[aria-label="Edit Shift Template"]')
      .simulate("save", shiftTemplate);
    expect(baseProps.addShiftTemplate).toBeCalled();
    expect(baseProps.addShiftTemplate).toBeCalledWith(shiftTemplate);
    expect(rotationPage.state("isCreatingOrEditingShiftTemplate")).toEqual(false);
  });

  it('should set isCreatingOrEditingShiftTemplate to false when closed', () => {
    const rotationPage = shallow(<RotationPage
      {...baseProps}
    />);
    rotationPage.setState({
      isCreatingOrEditingShiftTemplate: true
    });
    rotationPage.find('Connect(Component)[aria-label="Edit Shift Template"]').simulate("close");
    expect(rotationPage.state("isCreatingOrEditingShiftTemplate")).toEqual(false);
  });

  it('should call addShiftTemplate when a timeslot is selected', () => {
    const rotationPage = shallow(<RotationPage
      {...baseProps}
    />);
    const newDateStart = moment(baseDate).add(7, "days").toDate();
    const newDateEnd = moment(baseDate).add(7, "days").add(8, "hours").toDate();
    ((rotationPage.find('Schedule').props()) as { addEvent: Function}).addEvent(newDateStart,
      newDateEnd);

    expect(baseProps.addShiftTemplate).toBeCalled();
    expect(baseProps.addShiftTemplate).toBeCalledWith({
      tenantId: spot.tenantId,
      durationBetweenRotationStartAndTemplateStart: moment.duration(168, "hours"),
      shiftTemplateDuration: moment.duration(8, "hours"),
      spot: spot,
      rotationEmployee: null
    });
  });
});

const spot: Spot = {
  tenantId: 0,
  id: 2,
  version: 0,
  name: "Spot",
  requiredSkillSet: [
    {
      tenantId: 0,
      id: 3,
      version: 0,
      name: "Skill"
    }
  ]
}

const employee: Employee = {
  tenantId: 0,
  id: 4,
  version: 0,
  name: "Employee 1",
  contract: {
    tenantId: 0,
    id: 5,
    version: 0,
    name: "Basic Contract",
    maximumMinutesPerDay: 10,
    maximumMinutesPerWeek: 70,
    maximumMinutesPerMonth: 500,
    maximumMinutesPerYear: 6000
  },
  skillProficiencySet: [{
    tenantId: 0,
    id: 6,
    version: 0,
    name: "Not Required Skill"
  }]
}

const shiftTemplate: ShiftTemplate = {
  tenantId: 0,
  id: 1,
  version: 0, 
  durationBetweenRotationStartAndTemplateStart: moment.duration(1, "day").add(9, "hours"),
  shiftTemplateDuration: moment.duration(8, "hours"),
  spot: spot,
  rotationEmployee: employee
};

const rosterState: RosterState = {
  tenant: {
    id: 0,
    version: 0,
    name: "Tenant"
  },
  publishNotice: 14,
  publishLength: 7,
  firstDraftDate: new Date("2018-07-01"),
  draftLength: 7,
  unplannedRotationOffset: 0,
  rotationLength: 7,
  lastHistoricDate: new Date("2018-07-01"),
  timeZone: "EST"
};

const baseProps: Props & WithTranslation = {
  ...useTranslation(),
  tReady: true,
  isLoading: false,
  spotList: [spot],
  spotIdToShiftTemplateListMap: new Map<number, ShiftTemplate[]>([
    [2, [shiftTemplate]],
    [111, []]
  ]),
  rosterState: rosterState,
  addShiftTemplate: jest.fn(),
  removeShiftTemplate: jest.fn(),
  updateShiftTemplate: jest.fn(),
  showInfoMessage: jest.fn()
}