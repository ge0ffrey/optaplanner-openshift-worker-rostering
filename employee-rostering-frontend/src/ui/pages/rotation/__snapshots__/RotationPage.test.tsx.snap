// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rotation Page should render correctly when creating a new shift template via button 1`] = `
<Fragment>
  <Level
    className=""
    gutter="sm"
    style={
      Object {
        "backgroundColor": "var(--pf-global--BackgroundColor--100)",
        "height": "60px",
        "padding": "5px 5px 5px 5px",
      }
    }
  >
    <LevelItem
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <TypeaheadSelectInput
        aria-label="Select Spot"
        defaultValue={
          Object {
            "id": 2,
            "name": "Spot",
            "requiredSkillSet": Array [
              Object {
                "id": 3,
                "name": "Skill",
                "tenantId": 0,
                "version": 0,
              },
            ],
            "tenantId": 0,
            "version": 0,
          }
        }
        emptyText="Select Spot"
        onChange={[Function]}
        optionToStringMap={[Function]}
        options={
          Array [
            Object {
              "id": 2,
              "name": "Spot",
              "requiredSkillSet": Array [
                Object {
                  "id": 3,
                  "name": "Skill",
                  "tenantId": 0,
                  "version": 0,
                },
              ],
              "tenantId": 0,
              "version": 0,
            },
          ]
        }
      />
      <Pagination
        className=""
        dropDirection="down"
        itemCount={7}
        itemsEnd={null}
        itemsStart={null}
        onFirstClick={[Function]}
        onLastClick={[Function]}
        onNextClick={[Function]}
        onPageInput={[Function]}
        onPerPageSelect={[Function]}
        onPreviousClick={[Function]}
        onSetPage={[Function]}
        page={1}
        perPage={7}
        perPageOptions={Array []}
        titles={
          Object {
            "currPage": "Current week",
            "items": "days",
            "itemsPerPage": "Week #",
            "optionsToggle": "Select",
            "page": "week",
            "paginationTitle": "Week Select",
            "perPageSuffix": "Week #",
            "toFirstPage": "Go to the first week",
            "toLastPage": "Go to the last week",
            "toNextPage": "Go to the next week",
            "toPreviousPage": "Go to the previous week",
          }
        }
        toggleTemplate={[Function]}
        variant="top"
        widgetId="pagination-options-menu"
      />
    </LevelItem>
    <LevelItem
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <Button
        aria-label="Create Shift Template"
        onClick={[Function]}
        style={
          Object {
            "margin": "5px",
          }
        }
      >
        Create Shift Template
      </Button>
    </LevelItem>
  </Level>
  <Connect(Component)
    aria-label="Edit Shift Template"
    isOpen={true}
    onClose={[Function]}
    onDelete={[Function]}
    onSave={[Function]}
  />
  <Schedule
    addEvent={[Function]}
    dateFormat={[Function]}
    dayStyle={[Function]}
    endAccessor={[Function]}
    endDate={2018-01-07T00:00:00.000Z}
    eventComponent={[Function]}
    eventStyle={[Function]}
    events={
      Array [
        Object {
          "end": 2018-01-01T17:00:00.000Z,
          "shiftTemplate": Object {
            "durationBetweenRotationStartAndTemplateStart": "P1DT9H",
            "id": 1,
            "rotationEmployee": Object {
              "contract": Object {
                "id": 5,
                "maximumMinutesPerDay": 10,
                "maximumMinutesPerMonth": 500,
                "maximumMinutesPerWeek": 70,
                "maximumMinutesPerYear": 6000,
                "name": "Basic Contract",
                "tenantId": 0,
                "version": 0,
              },
              "id": 4,
              "name": "Employee 1",
              "skillProficiencySet": Array [
                Object {
                  "id": 6,
                  "name": "Not Required Skill",
                  "tenantId": 0,
                  "version": 0,
                },
              ],
              "tenantId": 0,
              "version": 0,
            },
            "shiftTemplateDuration": "PT8H",
            "spot": Object {
              "id": 2,
              "name": "Spot",
              "requiredSkillSet": Array [
                Object {
                  "id": 3,
                  "name": "Skill",
                  "tenantId": 0,
                  "version": 0,
                },
              ],
              "tenantId": 0,
              "version": 0,
            },
            "tenantId": 0,
            "version": 0,
          },
          "start": 2018-01-01T09:00:00.000Z,
        },
      ]
    }
    key="2"
    popoverBody={[Function]}
    popoverHeader={[Function]}
    startAccessor={[Function]}
    startDate={2017-12-31T00:00:00.000Z}
    titleAccessor={[Function]}
    wrapperStyle={[Function]}
  />
</Fragment>
`;

exports[`Rotation Page should render correctly when loaded 1`] = `
<Fragment>
  <Level
    className=""
    gutter="sm"
    style={
      Object {
        "backgroundColor": "var(--pf-global--BackgroundColor--100)",
        "height": "60px",
        "padding": "5px 5px 5px 5px",
      }
    }
  >
    <LevelItem
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <TypeaheadSelectInput
        aria-label="Select Spot"
        defaultValue={
          Object {
            "id": 2,
            "name": "Spot",
            "requiredSkillSet": Array [
              Object {
                "id": 3,
                "name": "Skill",
                "tenantId": 0,
                "version": 0,
              },
            ],
            "tenantId": 0,
            "version": 0,
          }
        }
        emptyText="Select Spot"
        onChange={[Function]}
        optionToStringMap={[Function]}
        options={
          Array [
            Object {
              "id": 2,
              "name": "Spot",
              "requiredSkillSet": Array [
                Object {
                  "id": 3,
                  "name": "Skill",
                  "tenantId": 0,
                  "version": 0,
                },
              ],
              "tenantId": 0,
              "version": 0,
            },
          ]
        }
      />
      <Pagination
        className=""
        dropDirection="down"
        itemCount={7}
        itemsEnd={null}
        itemsStart={null}
        onFirstClick={[Function]}
        onLastClick={[Function]}
        onNextClick={[Function]}
        onPageInput={[Function]}
        onPerPageSelect={[Function]}
        onPreviousClick={[Function]}
        onSetPage={[Function]}
        page={1}
        perPage={7}
        perPageOptions={Array []}
        titles={
          Object {
            "currPage": "Current week",
            "items": "days",
            "itemsPerPage": "Week #",
            "optionsToggle": "Select",
            "page": "week",
            "paginationTitle": "Week Select",
            "perPageSuffix": "Week #",
            "toFirstPage": "Go to the first week",
            "toLastPage": "Go to the last week",
            "toNextPage": "Go to the next week",
            "toPreviousPage": "Go to the previous week",
          }
        }
        toggleTemplate={[Function]}
        variant="top"
        widgetId="pagination-options-menu"
      />
    </LevelItem>
    <LevelItem
      style={
        Object {
          "display": "flex",
        }
      }
    >
      <Button
        aria-label="Create Shift Template"
        onClick={[Function]}
        style={
          Object {
            "margin": "5px",
          }
        }
      >
        Create Shift Template
      </Button>
    </LevelItem>
  </Level>
  <Connect(Component)
    aria-label="Edit Shift Template"
    isOpen={false}
    onClose={[Function]}
    onDelete={[Function]}
    onSave={[Function]}
  />
  <Schedule
    addEvent={[Function]}
    dateFormat={[Function]}
    dayStyle={[Function]}
    endAccessor={[Function]}
    endDate={2018-01-07T00:00:00.000Z}
    eventComponent={[Function]}
    eventStyle={[Function]}
    events={
      Array [
        Object {
          "end": 2018-01-01T17:00:00.000Z,
          "shiftTemplate": Object {
            "durationBetweenRotationStartAndTemplateStart": "P1DT9H",
            "id": 1,
            "rotationEmployee": Object {
              "contract": Object {
                "id": 5,
                "maximumMinutesPerDay": 10,
                "maximumMinutesPerMonth": 500,
                "maximumMinutesPerWeek": 70,
                "maximumMinutesPerYear": 6000,
                "name": "Basic Contract",
                "tenantId": 0,
                "version": 0,
              },
              "id": 4,
              "name": "Employee 1",
              "skillProficiencySet": Array [
                Object {
                  "id": 6,
                  "name": "Not Required Skill",
                  "tenantId": 0,
                  "version": 0,
                },
              ],
              "tenantId": 0,
              "version": 0,
            },
            "shiftTemplateDuration": "PT8H",
            "spot": Object {
              "id": 2,
              "name": "Spot",
              "requiredSkillSet": Array [
                Object {
                  "id": 3,
                  "name": "Skill",
                  "tenantId": 0,
                  "version": 0,
                },
              ],
              "tenantId": 0,
              "version": 0,
            },
            "tenantId": 0,
            "version": 0,
          },
          "start": 2018-01-01T09:00:00.000Z,
        },
      ]
    }
    key="2"
    popoverBody={[Function]}
    popoverHeader={[Function]}
    startAccessor={[Function]}
    startDate={2017-12-31T00:00:00.000Z}
    titleAccessor={[Function]}
    wrapperStyle={[Function]}
  />
</Fragment>
`;

exports[`Rotation Page should render correctly when loading 1`] = `<div />`;
